### PR TITLE
Use OneHotHashVectorizer instead of ToKey in sample 2.5.

### PR DESCRIPTION
**Pull request by pieths**
Fixes [NimbusML issue 59](https://github.com/microsoft/NimbusML/issues/59).

### CODE REVIEW

**Comment by najeeb-kazmi** [comment]
>        "0     0.2     0.2  0.228608  4.328085  7.058803  9.222845"

Was this cell run again with the ranker trained on the MSLR 10K dataset? I remember NDCG being in the ballpark of 0.45 on that dataset.

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:384 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by najeeb-kazmi** [comment]
>     "The Normalized Discounted Cumulative Gain (NDCG) and Discounted Cumulative Gain (DCG) at a particular rank position (1,2, or 3) are reported by the .test() function if setting evaltype='ranking'. DCG evaluates the quality of a rank with respect to the original relevance scores. It is the sum of the relevance scores for the top N items in the ranking, and it can be computed as follows: \n",

It is the sum of the discounted relevance scores for the top N items in the ranking. The discounting factor, which is logarithmically proportional to the position of the item, is meant to penalize highly relevant items that appear lower in the ranking. DCG can be computed as follows:

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:320 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by najeeb-kazmi** [comment]
>     "Where N is the threshold of ranking to account for in DCG, and g<sub>i</sub> the relavance score for the i<sup>th</sup> item in the ranking. The NDCG is simply calculated by normalizing the DCG by Max_DCG, which is the DCG for the perfect ranking (all the items are sorted by relevance score)."

Since the number of items in a result can vary by query, DCG is not a consistent metric when comparing results for multiple queries. NDCG normalizes DCG by the max(DCG) or the Ideal DCG (IDCG), which is the DCG for the perfect ranking where all items are ranked in decreasing order of relevance. This scales NDCG to a range of [0,1] so that comparisons of rankings across multiple queries are possible.

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:324 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by najeeb-kazmi** [comment]
>     "# scores, metrics = model.test(test_stream, evaltype='ranking') #Error, but I would expect this to work as model.fit()\n",

Remove this commented out line? 

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:299 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by najeeb-kazmi** [comment]
>     "scores[0:10]"

Maybe add another cell that shows a ranker's predictions in action? Something like:

scores[scores.QueryID == 13].sort_values('Score", ascending=False).head()

and a discussion about looking at the predicted scores for a particular query (e.g. query 13). The predicted ranking of items is made by sorting all the items for a particular query by score, with the highest score indicating the most relevant predicted item. We show the top few predicted results for this query here.

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:306 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by pieths** [comment]
>        "0     0.2     0.2  0.228608  4.328085  7.058803  9.222845"

I did not update the sample to use the real dataset due to licensing issues so the NDCG has not changed.

---
In reply to: [550066663](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550066663) [](ancestors = 550066663)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:384 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by pieths** [comment]
>     "The Normalized Discounted Cumulative Gain (NDCG) and Discounted Cumulative Gain (DCG) at a particular rank position (1,2, or 3) are reported by the .test() function if setting evaltype='ranking'. DCG evaluates the quality of a rank with respect to the original relevance scores. It is the sum of the relevance scores for the top N items in the ranking, and it can be computed as follows: \n",

The text has been updated.

---
In reply to: [550068487](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550068487) [](ancestors = 550068487)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:320 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by pieths** [comment]
>     "Where N is the threshold of ranking to account for in DCG, and g<sub>i</sub> the relavance score for the i<sup>th</sup> item in the ranking. The NDCG is simply calculated by normalizing the DCG by Max_DCG, which is the DCG for the perfect ranking (all the items are sorted by relevance score)."

The text has been updated.

---
In reply to: [550070609](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550070609) [](ancestors = 550070609)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:324 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by pieths** [comment]
>     "# scores, metrics = model.test(test_stream, evaltype='ranking') #Error, but I would expect this to work as model.fit()\n",

Removed.

---
In reply to: [550071092](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550071092) [](ancestors = 550071092)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:299 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by pieths** [comment]
>     "scores[0:10]"

The output had been updated to show the top 10 results.

---
In reply to: [550072868](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550072868) [](ancestors = 550072868)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:306 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)

**Comment by najeeb-kazmi** [comment]
>     "scores[0:10]"

Please add a comment above this line explaining what we are doing here, i.e. showing the top ten predicted items for query 5, where predictions are done by sorting on the predicted score within a group id.

---
In reply to: [550434532](https://github.com/microsoft/NimbusML-Samples/pull/30#issuecomment-550434532) [](ancestors = 550434532,550072868)

---
Refers to: samples/2.5 [Numeric] Learning-to-Rank with Microsoft Bing Data.ipynb:306 in e8993d7. [](commit_id = e8993d73d2a5444fa8706bd0c427fc9df6e9997c, deletion_comment = False)
